packages/flutter_tools/lib/src/test/runner.dart

```diff
@@ -60,49 +154,85 @@ Future<int> runTests(
-    precompiledDillFiles: precompiledDillFiles,
-    trackWidgetCreation: trackWidgetCreation,
-    updateGoldens: updateGoldens,
-    projectRootDirectory: fs.currentDirectory.uri,
-  );
+  return testArgs;
+}
 
-  // Make the global packages path absolute.
-  // (Makes sure it still works after we change the current directory.)
-  PackageMap.globalPackagesPath =
-      fs.path.normalize(fs.path.absolute(PackageMap.globalPackagesPath));
+typedef _HookInstaller = void Function(Map<String, String> precompiledDillFiles, Map<String, List<Finalizer>> finalizers);
 
-  // Call package:test's main method in the appropriate directory.
-  final Directory saved = fs.currentDirectory;
-  try {
-    if (workDir != null) {
-      printTrace('switching to directory $workDir to run tests');
-      fs.currentDirectory = workDir;
+Future<void> _compileTestFiles(
+  TestCompiler compiler,
+  List<String> testFiles, {
+  _HookInstaller installHook,
+  InternetAddress host,
+  bool updateGoldens,
+  List<String> invalidatedFiles = const <String>[]
+}) async {
+  int index = 0;
+  final Map<String, String> precompiledDillFiles = <String, String>{};
+  final Map<String, List<Finalizer>> finalizers = <String, List<Finalizer>>{};
+  for (String file in testFiles) {
+    final List<Finalizer> fileFinalizers = <Finalizer>[];
+    final String mainDart = createListenerDart(
+      fileFinalizers,
+      index,
+      file,
+      host,
+      updateGoldens,
+    );
+
+    final String dillPath = await compiler.compile(mainDart, invalidatedFiles: invalidatedFiles);
+    // We only need to invalidate the changed file once, so clear the list
+    if (invalidatedFiles.isNotEmpty) {
+      invalidatedFiles.clear();
     }
+    precompiledDillFiles[file] = dillPath;
+    finalizers[file] = fileFinalizers;
+    index++;
+  }
 
-    printTrace('running test package with arguments: $testArgs');
-    await test.main(testArgs);
+  installHook(precompiledDillFiles, finalizers);
+}
 
-    // test.main() sets dart:io's exitCode global.
-    printTrace('test package returned with exit code $exitCode');
+void _printReloadMessage() {
+  printStatus(
+    "Press 'r' to rerun your tests, 'q' to quit",
+    color: TerminalColor.red
+  );
+}
 
-    return exitCode;
-  } finally {
-    fs.currentDirectory = saved;
+Future<void> _testWatch({
+  Directory testDir,
+  List<String> testArgs,
+  List<String> testFiles,
+  TestCompiler compiler,
+  InternetAddress host,
+  bool updateGoldens,
+  _HookInstaller installHook
+}) async {
+  final List<String> invalidatedFiles = <String>[];
+  final DirectoryWatcher directoryWatcher = DirectoryWatcher(testDir.path);
+
+  directoryWatcher.events.listen((WatchEvent event) {
+    if (event.path.endsWith('.dart')) {
+      invalidatedFiles.add(event.path);
+    }
+  });
+  _printReloadMessage();
+
+  await directoryWatcher.ready;
+  terminal.singleCharMode = true;
+  await for (String char in terminal.keystrokes.asBroadcastStream()){
```

Comment:
Note that holding down `r` queues up multiple runs. You will probably need some sort of "busy" signal to ignore the stdin while you are running a test. Example
```
bool isBusy = false;

terminal.keystrokes. asBroadcastStream().where((x) =>!isBusy);
...
isBusy = true;
await _compileTestFiles(...);
isBusy = false;
```

Reply:
Great catch.  Done.